# Python and R scripts

Python and R code (some used, some not) from the project

For help email: k.henry@dkfz-heidelberg.de or brookshenry3@gmail.com

## Resources:

 * [NCSU LSF Template help page](https://projects.ncsu.edu/hpc/Documents/lsf_template.php)
 * [NCSU Running jobs with LSF](https://projects.ncsu.edu/hpc/Documents/LSF.php)
 * [IBM LSF documentation](https://www.ibm.com/support/knowledgecenter/en/SSWRJV_10.1.0/lsf_welcome/lsf_welcome.html)

## Description of files:

 * alignment_correlation.py: 
 * bam_to_fasta.py: convert bam file to fasta file
 * do_canu.sh: [Canu de novo assembler](https://github.com/marbl/canu)
 * do_clair.sh: [Clair SV caller](https://github.com/HKU-BAL/Clair)
 * do_concensus.sh: Create a concensus vcf file from a bam file
 * do_cutesv.sh: [CuteSV SV caller](https://github.com/tjiangHIT/cuteSV)
 * do_f5c.sh: [f5c methylation caller](https://github.com/hasindu2008/f5c)
 * do_flye.sh: [Flye de novo assembler](https://github.com/fenderglass/Flye)
 * do_guppy_gpu.sh: Guppy basecaller - GPU enabled version (available on ONT community site)
 * do_guppy.sh: Guppy basecaller - CPU version (available on ONT community site)
 * do_gzip.sh: gzip a merged fastq file
 * do_gztobgz.sh: Convert a gzipped file into a bgzipped file
 * do_index.sh: [Samtools index](https://github.com/samtools/samtools)
 * do_merge.sh: merge multiple fastq files into a single fastq file
 * do_minimap2.sh: [minimap2 alignment tool](https://github.com/lh3/minimap2)
 * do_ngmlr.sh: [NGMLR alignment tool](https://github.com/philres/ngmlr)
 * do_pileup.sh: Generate a pileup file from a bam file
 * do_pycoqc.sh: [PycoQC sequencing quality assessment tool](https://github.com/a-slide/pycoQC)
 * do_slice_bam.sh: Parse a bam file for a region of interest
 * do_sniffles.sh: [Sniffles SV caller](https://github.com/fritzsedlazeck/Sniffles)
 * do_wengan.sh: [WENGAN de novo assembler](https://github.com/adigenova/wengan)
 

 