
SAMPLES = ['AR']
#INDEX = "/icgc/dkfzlsdf/analysis/C010/brooks/output/ashish"

#Imports
import os


#configfile: "config.yaml"


#SAMPLES = config['SAMPLES']
#REF = config['REF']

#Activating virtual environment, loading needed tools
def activate_environment():
    os.system("module load samtools/1.9")
    os.system("module load anaconda3/2019.07")
    os.system("source activate master-env")

activate_environment()

#Expected output
rule all:
    input:
        #expand("/icgc/dkfzlsdf/analysis/C010/brooks/output/GCTB/{sample}_pycoQC_output.html", sample=SAMPLES),
        #expand("/icgc/dkfzlsdf/analysis/C010/brooks/output/GCTB/{sample}.merged.fastq", sample=SAMPLES),
        #expand("/icgc/dkfzlsdf/analysis/C010/brooks/output/GCTB/{sample}.merged.fastq.index.gzi", sample=SAMPLES),
        #expand("/icgc/dkfzlsdf/analysis/C010/brooks/output/GCTB/{sample}.merged.fastq.index.fai", sample=SAMPLES),
        #expand("/icgc/dkfzlsdf/analysis/C010/brooks/output/GCTB/{sample}.merged.fastq.index.readdb", sample=SAMPLES),
        expand("output/GCTB/{sample}.sorted.bam", sample=SAMPLES),
        expand("output/GCTB/{sample}.sorted.bam.bai", sample=SAMPLES),
        expand("output/GCTB/{sample}.meth_calls.tsv", sample=SAMPLES),
        expand("output/GCTB/{sample}.meth_freq.tsv", sample=SAMPLES),
        expand("output/GCTB/{sample}.vcf", sample=SAMPLES),
        expand("output/GCTB/{sample}.methrix.bedGraph", sample=SAMPLES)

#rule pycoqc:
#    input:
#        "/icgc/dkfzlsdf/analysis/C010/brooks/nanopore_basecalled/{sample}/hac_cfg/sequencing_summary.txt"
#    output:
#        "/icgc/dkfzlsdf/analysis/C010/brooks/output/GCTB/{sample}_pycoQC_output.html"
#    params:
#        jobname="QC",
#        runtime="4:00",
#        memusage="5000",
#        slots="1",
#        misc=" "
#    shell:
#        "pycoQC -f {input} -o {output}"

#rule merge_fastqs:
#    input:
#        "/icgc/dkfzlsdf/analysis/C010/brooks/nanopore_basecalled/{sample}/hac_cfg/"
#    output:
#        "/icgc/dkfzlsdf/analysis/C010/brooks/output/GCTB/{sample}.merged.fastq"
#    params:
#        jobname="merge_fqs",
#        runtime="5:00",
#        memusage="5000",
#        slots="1",
#        misc=" "
#    shell:
#        "cat {input}/*.fastq > {output}"

#rule index:
#    input:
#        f5s="/icgc/dkfzlsdf/analysis/C010/brooks/nanopore_raw/{sample}/fast5/", 
#        fq="/icgc/dkfzlsdf/analysis/C010/brooks/output/GCTB/{sample}.merged.fastq"
#    output:
#        "output/GCTB/{sample}.merged.fastq.index",
#        "output/GCTB/{sample}.merged.fastq.index.gzi",
#        "output/GCTB/{sample}.merged.fastq.index.fai",
#        "output/GCTB/{sample}.merged.fastq.index.readdb"
#    params:
#        jobname="index_reads",
#        runtime="12:00",
#        memusage="5000",
#        slots="10",
#        misc=" "
#    shell:
#        "f5c index -d {input.f5s} {input.fq}"

rule minimap2_align:
    input:
        ref="/icgc/dkfzlsdf/analysis/C010/brooks/GCF_000001405.26_GRCh38_genomic.fna",
        fq="output/GCTB/{sample}.merged.fastq"
    output:
        "output/GCTB/{sample}.sorted.bam"
    params:
        jobname="index_reads",
        runtime="8:00",
        memusage="20000",
        slots="10",
        misc=" "
    shell:
        "minimap2 -t {params.slots} -ax map-ont {input.ref} {input.fq} | samtools sort -T {wildcards.sample} -o {output}"

rule index_bams:
    input:
        "output/GCTB/{sample}.sorted.bam"
    output:
        "output/GCTB/{sample}.sorted.bam.bai"
    params:
        jobname="index_bams",
        runtime="8:00",
        memusage="20000",
        slots="1",
        misc=" "
    shell:
        "samtools index {input}"

rule call_meth:
    input:
        bam="output/GCTB/{sample}.sorted.bam",
        bai="output/GCTB/{sample}.sorted.bam.bai",
        ref="/icgc/dkfzlsdf/analysis/C010/brooks/GCF_000001405.26_GRCh38_genomic.fna",
        fq="output/GCTB/{sample}.merged.fastq",
        index="output/GCTB/{sample}.merged.fastq.index",
        fai="output/GCTB/{sample}.merged.fastq.index.fai",
        gzi="output/GCTB/{sample}.merged.fastq.index.gzi",
        readdb="output/GCTB/{sample}.merged.fastq.index.readdb"
    output:
        "output/GCTB/{sample}.meth_calls.tsv"
    params:
        jobname="call_meth",
        runtime="48:00",
        memusage="10000",
        slots="10",
        misc=" "
    shell:
        "f5c call-methylation -t 10 -r {input.fq} -b {input.bam} -g {input.ref} > {output}"

rule calc_freq:
    input:
        "output/GCTB/{sample}.meth_calls.tsv"
    output:
        "output/GCTB/{sample}.meth_freq.tsv"
    params:
        jobname="calc_freq",
        runtime="8:00",
        memusage="5000",
        slots="1",
        misc=" "
    shell:
        "f5c meth-freq -i {input} -o {output} -s"

rule call_vars:
    input:
        bam="output/GCTB/{sample}.sorted.bam",
        ref="/icgc/dkfzlsdf/analysis/C010/brooks/GCF_000001405.26_GRCh38_genomic.fna" #,
        #wd="/icgc/dkfzlsdf/analysis/C010/brooks/master_pipeline/output/{sample}"
    output:
        "output/GCTB/{sample}.vcf"
    params:
        jobname="call_vars",
        runtime="8:00",
        memusage="10000",
        slots="20",
        misc=" "
    shell:
        """
        mkdir /icgc/dkfzlsdf/analysis/C010/brooks/master_pipeline/output/GCTB/{wildcards.sample}
        cuteSV {input.bam} {input.ref} {output} /icgc/dkfzlsdf/analysis/C010/brooks/master_pipeline/output/GCTB/{wildcards.sample} \
        --threads {params.slots} --max_cluster_bias_INS 100 --diff_ratio_merging_INS 0.3 --max_cluster_bias_DEL 100 --diff_ratio_merging_DEL 0.3
        
        """
#Might want to add in rm -r part to rule above to get rid of empty folders 

rule tsv_to_bg:
    input:
        "output/GCTB/{sample}.meth_freq.tsv"
    output:
        "output/GCTB/{sample}.methrix.bedGraph"
    params:
        jobname='frequency_to_methrix',
        runtime='30:00',
        memusage='10000',
        slots='2',
        misc=''
    shell:
        "python3 reformat_tsv.py -f {input} -o {output}"